File: stripe/src/test/java/com/stripe/android/model/StripeSourceTypeModelTest.java
package com.stripe.android.model;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import static com.stripe.android.model.StripeSourceTypeModel.jsonObjectToMapWithoutKeys;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

/**
 * Test class for {@link StripeSourceTypeModel}.
 */
public class StripeSourceTypeModelTest {

    @Test
    public void jsonObjectToMapWithoutKeys_whenHasKeyInput_returnsMapOmittingKeys()
            throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("a_key", "a_value");
        jsonObject.put("b_key", "b_value");
        jsonObject.put("c_key", "c_value");
        jsonObject.put("d_key", "d_value");

        Set<String> omitKeys = new HashSet<String>() {{
            add("a_key");
            add("d_key");
        }};
        Map<String, Object> resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys);
        assertNotNull(resultMap);
        assertEquals(2, resultMap.size());
        assertEquals("b_value", resultMap.get("b_key"));
        assertEquals("c_value", resultMap.get("c_key"));
        assertFalse(resultMap.containsKey("a_key"));
        assertFalse(resultMap.containsKey("d_key"));
    }

    @Test
    public void jsonObjectToMapWithoutKeys_whenAllKeysGiven_returnsNull() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("a_key", "a_value");
        jsonObject.put("b_key", "b_value");

        Set<String> omitKeys = new HashSet<String>() {{
            add("a_key");
            add("b_key");
        }};
        Map<String, Object> resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys);
        assertNull(resultMap);
    }

    @Test
    public void jsonObjectToMapWithoutKeys_whenOtherKeysGiven_returnsFullMap() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("a_key", "a_value");
        jsonObject.put("b_key", "b_value");

        Set<String> omitKeys = new HashSet<String>() {{
            add("c_key");
            add("d_key");
        }};
        Map<String, Object> resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys);
        assertNotNull(resultMap);
        assertEquals("a_value", resultMap.get("a_key"));
        assertEquals("b_value", resultMap.get("b_key"));
    }
}

File: stripe/src/test/java/com/stripe/android/model/StripeSourceTypeModelTest.kt
package com.stripe.android.model

import com.stripe.android.model.StripeSourceTypeModel.Companion.jsonObjectToMapWithoutKeys
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import org.json.JSONObject

/**
 * Test class for [StripeSourceTypeModel].
 */
class StripeSourceTypeModelTest {

    @Test
    fun jsonObjectToMapWithoutKeys_whenHasKeyInput_returnsMapOmittingKeys() {
        val jsonObject = JSONObject()
            .put("a_key", "a_value")
            .put("b_key", "b_value")
            .put("c_key", "c_value")
            .put("d_key", "d_value")

        val omitKeys = setOf("a_key", "d_key")

        val resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys)!!
        assertEquals(2, resultMap.size)
        assertEquals("b_value", resultMap["b_key"])
        assertEquals("c_value", resultMap["c_key"])
        assertFalse(resultMap.containsKey("a_key"))
        assertFalse(resultMap.containsKey("d_key"))
    }

    @Test
    fun jsonObjectToMapWithoutKeys_whenAllKeysGiven_returnsNull() {
        val jsonObject = JSONObject()
            .put("a_key", "a_value")
            .put("b_key", "b_value")

        val omitKeys = setOf("a_key", "b_key")
        val resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys)
        assertNull(resultMap)
    }

    @Test
    fun jsonObjectToMapWithoutKeys_whenOtherKeysGiven_returnsFullMap() {
        val jsonObject = JSONObject()
            .put("a_key", "a_value")
            .put("b_key", "b_value")

        val omitKeys = setOf("c_key", "d_key")
        val resultMap = jsonObjectToMapWithoutKeys(jsonObject, omitKeys)!!
        assertEquals("a_value", resultMap["a_key"])
        assertEquals("b_value", resultMap["b_key"])
    }
}
